### PR TITLE
feat: Add a management command to bulk dump course data to ClickHouse

### PR DESCRIPTION
**Pull request by bmtcril**
**Description:** Adds a Django management command for bulk dumping course data to ClickHouse.

**ISSUE:** #5 

**Testing instructions:**

1. In an otherwise working OARS env)
2. Install the OARS plugin branc: https://github.com/openedx/tutor-contrib-oars/pull/35
3. `tutor config save` (to get the ClickHouse config)
4. `tutor images build openedx --no-cache` (to get the plugin installed)
5. `tutor local do init -l oars` (to get the new database and tables created)
6. `tutor local start`
7. Update content in a course and check the "cms worker logs" to see that real-ish time course published sink still functions
8. Shell into the CMS or CMS worker container and experiment with the following commands to see courses dumped to Clickhouse in bulk:

Help: `python manage.py cms dump_courses_to_clickhouse -h`
Dump all eligible courses: `python manage.py cms dump_courses_to_clickhouse`
Dump all courses, no matter their last dump time: `python manage.py cms dump_courses_to_clickhouse --force`
Dump up to 1 course: `python manage.py cms dump_courses_to_clickhouse --limit 1`
Dump to a different ClickHouse than the one in settings: `python manage.py cms dump_courses_to_clickhouse --url https://foo.bar`

You should see updated course data in ClickHouse's event_sink.course_blocks and event_sink.course_relationships

**Reviewers:**
- [ ] @pomegranited 
- [ ] @Ian2012 
- [ ] @mariajgrimaldi

**Merge checklist:**
- [ ] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [x] Documentation updated (not only docstrings)
- [x] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPI after tag-triggered build is
      finished.
- [ ] Delete working branch (if not needed anymore)

### CODE REVIEW

**Comment by Ian2012** [comment]
@bmtcril Data is replicated every time with the `--force` flag, Should tables be deleted before proceeding with the dump?

**Review by Ian2012 — APPROVED** [review]
LGTM

**Comment by bmtcril** [comment]
@Ian2012 it's a good catch about the --force thing. I think that we'll need to use the last dump time / dump id to disambiguate those. Deleting from ClickHouse on these kinds of tables can cause some serious performance issues, but with that other metadata I think it's a thing we can handle in dbt.

**Comment by mariajgrimaldi** [comment]
I was testing the command with the latest changes and found that: 

```
cms-worker_1                 | [2023-05-10 22:11:07,357: ERROR/ForkPoolWorker-7] Code: 60. DB::Exception: Table event_sink.course_overviews doesn't exist. (UNKNOWN_TABLE) (version 23.3.2.37 (official build))
cms-worker_1                 | 
cms-worker_1                 | [2023-05-10 22:11:07,357: ERROR/ForkPoolWorker-7] Error trying to dump course course-v1:edX+DemoX+Demo_Course to ClickHouse!
cms-worker_1                 | Traceback (most recent call last):
cms-worker_1                 |   File "/openedx/venv/lib/python3.8/site-packages/event_sink_clickhouse/sinks/course_published.py", line 306, in dump
cms-worker_1                 |     self._send_course_overview(serialized_courseoverview)
cms-worker_1                 |   File "/openedx/venv/lib/python3.8/site-packages/event_sink_clickhouse/sinks/course_published.py", line 239, in _send_course_overview
cms-worker_1                 |     self._send_clickhouse_request(request)
cms-worker_1                 |   File "/openedx/venv/lib/python3.8/site-packages/event_sink_clickhouse/sinks/base_sink.py", line 46, in _send_clickhouse_request
cms-worker_1                 |     response.raise_for_status()
cms-worker_1                 |   File "/openedx/venv/lib/python3.8/site-packages/requests/models.py", line 1021, in raise_for_status
cms-worker_1                 |     raise HTTPError(http_error_msg, response=self)
cms-worker_1                 | requests.exceptions.HTTPError: 404 Client Error: Not Found for url: http://clickhouse:8123/?input_format_allow_errors_num=1&input_format_allow_errors_ratio=0.1&query=INSERT+INTO+event_sink.course_overviews+FORMAT+CSV
cms-worker_1                 | [2023-05-10 22:11:07,359: ERROR/ForkPoolWorker-7] Task event_sink_clickhouse.tasks.dump_course_to_clickhouse[e76391ae-36cc-45f8-a6a5-b3d50376489a] raised unexpected: HTTPError('404 Client Error: Not Found for url: http://clickhouse:8123/?input_format_allow_errors_num=1&input_format_allow_errors_ratio=0.1&query=INSERT+INTO+event_sink.course_overviews+FORMAT+CSV')
cms-worker_1                 | Traceback (most recent call last):
cms-worker_1                 |   File "/openedx/venv/lib/python3.8/site-packages/celery/app/trace.py", line 451, in trace_task
cms-worker_1                 |     R = retval = fun(*args, **kwargs)
cms-worker_1                 |   File "/openedx/venv/lib/python3.8/site-packages/celery/app/trace.py", line 734, in __protected_call__
cms-worker_1                 |     return self.run(*args, **kwargs)
cms-worker_1                 |   File "/openedx/venv/lib/python3.8/site-packages/edx_django_utils/monitoring/internal/code_owner/utils.py", line 193, in new_function
cms-worker_1                 |     return wrapped_function(*args, **kwargs)
cms-worker_1                 |   File "/openedx/venv/lib/python3.8/site-packages/event_sink_clickhouse/tasks.py", line 30, in dump_course_to_clickhouse
cms-worker_1                 |     sink.dump(course_key)
cms-worker_1                 |   File "/openedx/venv/lib/python3.8/site-packages/event_sink_clickhouse/sinks/course_published.py", line 306, in dump
cms-worker_1                 |     self._send_course_overview(serialized_courseoverview)
cms-worker_1                 |   File "/openedx/venv/lib/python3.8/site-packages/event_sink_clickhouse/sinks/course_published.py", line 239, in _send_course_overview
cms-worker_1                 |     self._send_clickhouse_request(request)
cms-worker_1                 |   File "/openedx/venv/lib/python3.8/site-packages/event_sink_clickhouse/sinks/base_sink.py", line 46, in _send_clickhouse_request
cms-worker_1                 |     response.raise_for_status()
cms-worker_1                 |   File "/openedx/venv/lib/python3.8/site-packages/requests/models.py", line 1021, in raise_for_status
cms-worker_1                 |     raise HTTPError(http_error_msg, response=self)
cms-worker_1                 | requests.exceptions.HTTPError: 404 Client Error: Not Found for url: http://clickhouse:8123/?input_format_allow_errors_num=1&input_format_allow_errors_ratio=0.1&query=INSERT+INTO+event_sink.course_overviews+FORMAT+CSV

```
Steps to reproduce:
- Follow testing instructions
- Publish a course
- Check the CMS worker logs

**Review by pomegranited — APPROVED** [review]
:+1: Works great for me @bmtcril !

Left a couple of minor nits, see inline comments.

- [x] I tested this I tested this with a fresh Tutor local devstack, and by following the PR instructions.
- [x] I read through the code
- [x] <strike>I checked for accessibility issues</strike> N/A
- [x] <strike>Includes documentation</strike> N/A

> `cms-worker_1                 | requests.exceptions.HTTPError: 404 Client Error: Not Found for url: http://clickhouse:8123/?input_format_allow_errors_num=1&input_format_allow_errors_ratio=0.1&query=INSERT+INTO+event_sink.course_overviews+FORMAT+CSV`

@mariajgrimaldi I didn't encounter the error you saw ^, but maybe because I rebuilt from scratch?

File: event_sink_clickhouse/management/commands/dump_courses_to_clickhouse.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        and one of courses that did not.
+    """
+    sink = CoursePublishedSink(connection_overrides, log)
+
+    submitted_courses = []
+    skipped_courses = []
+
+    index = 0
+    for course_key, should_be_dumped, reason in sink.fetch_target_courses(course_keys, courses_to_skip, force):
+        log.error(f"Iteration {index}")
```
[the remark this answers is not in the document]
Comment (bmtcril) [reply event_sink_clickhouse/management/commands/dump_courses_to_clickhouse.py:59]: Nope, I'll add the course key

File: event_sink_clickhouse/sinks/base_sink.py [hunk 2]
```diff
@@ -20,6 +20,7 @@ def __init__(self, connection_overrides, log):
                                       settings.EVENT_SINK_CLICKHOUSE_BACKEND_CONFIG["password"])
         self.ch_database = settings.EVENT_SINK_CLICKHOUSE_BACKEND_CONFIG["database"]
         self.ch_timeout_secs = settings.EVENT_SINK_CLICKHOUSE_BACKEND_CONFIG["timeout_secs"]
+        self.connection_overrides = connection_overrides
```
[the remark this answers is not in the document]
Comment (bmtcril) [reply event_sink_clickhouse/sinks/base_sink.py:24]: Seems like it's not, I'll remove it

**Comment by bmtcril** [comment]
@mariajgrimaldi the schema changed a few times over the course of trying to get things tested locally while github was on fire. If you drop everything in the event_sink ClickHouse database, pull the tutor-contrib-oars branch, and init oars again it should clear everything up for you. Sorry for the hassle.

@pomegranited @Ian2012 this is all repaired, squashed, and ready for re-review. I think all of the current feedback has been addressed.

File: event_sink_clickhouse/management/commands/dump_courses_to_clickhouse.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        parser.add_argument(
+            '--limit',
+            type=int,
+            help="maximum number of courses to dump, cannot be used with '--courses' or '--force'",
+        )
+
+    def handle(self, *args, **options):
+        """
+        Iterates through each course, serializes them into graphs, and saves
+        those graphs to neo4j.
```
Comment (Ian2012) [inline event_sink_clickhouse/management/commands/dump_courses_to_clickhouse.py:152]:
```suggestion
        those graphs to clickhouse.
```

**Comment by bmtcril** [comment]
@Ian2012 correct all of the data can and will be duplicated many times. Data in the xapi events corresponds to the version of the course that's live when the event is fired, so in some cases it's important to have the correct corresponding xblock data. We won't have full historical data, but this should let us be more accurate where we need to be (especially if problems change). dbt will be responsible for de-duping this in a way that's useful to each use case, but I can foresee there being different models for "most recent" and "accurate at emission time".

**Comment by bmtcril** [comment]
(also I'm not going to worry about that coverage at this point, this PR is holding up too many things)

**Review by Ian2012 — APPROVED** [review]

**Comment by bmtcril** [comment]
I've got two thumbs so I'm going to merge this. @mariajgrimaldi if you find other issues just let me know and I'll fix them in another PR.

**Pull request merged**